File: Assets/Scripts/DataChannel/DataChannel.cs
using UnityEngine;
using Unity.MLAgents;
using Unity.MLAgents.SideChannels;
using System;
using System.Collections.Generic;

public class DataChannel : SideChannel{
	static private Dictionary<string, int> intParameters = new Dictionary<string, int>();

	static public int getParameter(string key, int defaultValue){
		if(intParameters.ContainsKey(key))
			return intParameters[key];
		return defaultValue;
	}

	public DataChannel(){
		ChannelId = new Guid("9bc23f51-e0e8-450c-b3c5-e4d2032151ec");
	}

	protected override void OnMessageReceived(IncomingMessage msg){
		string receivedString = msg.ReadString();
	
		string[] segments = receivedString.Split('|');
		foreach(string seg in segments){
			Debug.Log(seg);
		}
		switch(segments[0]){
			case "int":
				intParameters[segments[1]] = int.Parse(segments[2]);
				break;
			default:
				Debug.LogError("Didn't recognize data type of message");
				break;
		}
	}
}

File: src/simulation/data_channel.py
import uuid

from mlagents_envs.side_channel.side_channel import IncomingMessage, OutgoingMessage, SideChannel


class DataChannel(SideChannel):
    def __init__(self) -> None:
        super().__init__(uuid.UUID("9bc23f51-e0e8-450c-b3c5-e4d2032151ec"))

    def on_message_received(self, msg: IncomingMessage) -> None:
        print(msg.read_string())

    def set_int_parameter(self, key: str, val: int) -> None:
        msg = OutgoingMessage()
        msg.write_string(f"int|{key}|{val}")
        super().queue_message_to_send(msg)

    def set_color_parameter(self, key: str, val: tuple[int, ...]) -> None:
        msg = OutgoingMessage()
        msg.write_string(f"color|{key}|{val[0]},{val[1]},{val[2]}")
        super().queue_message_to_send(msg)

    def set_float_parameter(self, key: str, val: float) -> None:
        msg = OutgoingMessage()
        msg.write_string(f"float|{key}|{val}")
        super().queue_message_to_send(msg)

    def set_bool_parameter(self, key: str, val: bool) -> None:
        msg = OutgoingMessage()
        msg.write_string(f"int|{key}|{1 if val else 0}")
        super().queue_message_to_send(msg)
